### DISCUSSION TITLE
Generic ORM

### DISCUSSION THREAD

**Discussion by Brayan-724** [RustLangES/.github#1] [Ideas]
> Sigue el desarrollo mas de cerca [aqui](https://discord.com/channels/778674594856960012/1270999369809788928)

La idea es generar un la abstraccion en traits genericos para que cada proyecto pueda usar las abstracciones para escribir su propio SQL

Algo como:

```rust
struct MyStruct {
  name: String,
  age: u8,
}

impl SqlExcecutor for MyStruct {
  type OkResponse = ();
  type Error = MyErrors;

  fn exec(&self) -> Result<OkResponse, Error> {
    let _ = db.exec("SELECT * FROM mytable")?;
    Ok(())
  }
}
```

**Comment by lsproule** [comment]
estos son locos

**Reply by lsproule** [reply]
pero son buena gente

**Comment by SergioRibera** [comment]
En base a lo que veníamos discutiendo en los canales de Voz del Discord, estuvimos pensando que incluso podríamos generar abstracciones desde el ORM para ciertos clientes como CloudFlare Workers (D1), AWS u otros

**Comment by romancitodev** [comment]
+1

**Comment by rebuild1157** [comment]
De hecho, yo hice algo así hace un tiempo. Fue algo complicado y probablemente poco óptimo, pero al menos me atrevería a decir que era compatible con una buena cantidad de interfaces de bases de datos y soportaba operaciones asíncronas.

**Reply by SergioRibera** [reply]
Tienes algun repositorio que podamos revisar?

Nosotros estuvimos trabajando en un proyecto en el que se tuvo que prototipar apresuradamente alguna solucion muy relacionada a la idea original de esta discusion ([aqui el proyecto](https://github.com/RustLangES/cfworker-forms/blob/apika/dev/crates/models/src/shared.rs)), el proyecto lo estuvo trabajando @Brayan-724 pero se alinea en lo que se busca con un ORM Generico

**Reply by rebuild1157** [reply]
[aquí está](https://github.com/ingrese1nombre/gnify/blob/main/crates/libs/base/src/source.rs). Lo he dejado abandonado desde hace rato, y creo que tiene una pequeña modificación pendiente. Pero en esencia es algo así. [aquí hay una implementación de la fuente de datos usando postgres con sqlx](https://github.com/ingrese1nombre/gnify/blob/main/crates/libs/base/src/source/postgres.rs#L29) (Disculpen el desorden) y [aquí algunos controladores para un modelo de usuario](https://github.com/ingrese1nombre/gnify/blob/main/crates/libs/core/src/user/bmc.rs)

**Reply by SergioRibera** [reply]
Genial, muchas gracias por los ejemplos, serán útiles para la prueba de concepto que vamos a estar desarrollando
